### DISCUSSION TITLE
Validating data in Setter in TemplatedControl

### DISCUSSION THREAD

**Discussion by BendRocks** [AvaloniaUI/Avalonia#14254] [Q&A]
New to Avalonia and I'm blown away with effort! I've seen the docs on doing data validation in the setter by throwing an exception ([HERE](https://docs.avaloniaui.net/docs/guides/development-guides/data-validation)), and it works as expected:

```
 public partial class MainWindowViewModel : ViewModelBase
 {
     public string Greeting => "Welcome to Avalonia!";

     public string _MyText = "1000";

     public string MyText
     {
         get { return _MyText; }

         set
         {
             if (double.TryParse(value, out double doubleVal) == false)
             {
                 throw new DataValidationException("Not a double");
             }
             _MyText = doubleVal.ToString();
         }
     }
 }
```

Based on the suggestion [HERE](https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-advanced-custom-controls), I've tried something similar in the codebehind for a templated control, but it crashes the entire app. 

```
public partial class TemplatedControl1 : TemplatedControl
{
    public static readonly DirectProperty<TemplatedControl1, string> MyTextProperty =
        AvaloniaProperty.RegisterDirect<TemplatedControl1, string>(
            nameof(MyText),
            getter: o => o.MyText,
            setter:(o, v) => o.MyText = v,
            enableDataValidation: true);

    private string _MyText;
    public string MyText
    {
        get
        {
            return _MyText;
        }
        set
        {
            // This line crashes the entire app
            //throw new DataValidationException("Not a double");
          
            SetAndRaise(MyTextProperty, ref _MyText, value);
        }
    }
```

The xaml for above is here:

```
<Style Selector="controls|TemplatedControl1">
	<!-- Set Defaults -->
	<Setter Property="Template">
		<ControlTemplate>
			<DataValidationErrors Errors="{TemplateBinding (DataValidationErrors.Errors)}">
				<TextBox Text="{TemplateBinding MyText}" Width="100" />
			</DataValidationErrors>
		</ControlTemplate>
	</Setter>
</Style>
```

My question is this: Is there is a way to do the validation in the setter of a property in a ```TemplatedControl```? If not, what is needed? Thanks!

**Comment by timunie** [comment]
Please check Avalonia.Samples > RatingControl. It has validation support included.

**Reply by BendRocks** [reply]
Hi @timunie, thanks for the pointer! I have spent some time RatingsControlSample, and converted to Community Toolkit. I think the summary is that exceptions in the setter wont' work for flagging invalid data in a templated control. At best, it appears you need to bind to the property in the templatedcontrol with a property in a ViewModel (as the RatingsControlSample does). Once you have done that, then you can override UpdateDataValidation in the templatedcontrol and validate the data there if needed (or attach a constraint to the property in the ViewModel as the RatingControl has done. 

But without the binding to the ViewModel propety, the UpdateDataValidation never fires.

**Reply by timunie** [reply]
Think you can simply call 

```
ValidationErrors.SetError 
```

Anywhere. For example in property change override
